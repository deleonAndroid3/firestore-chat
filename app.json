{
  "expo": {
    "name": "Firestore Chat",
    "slug": "firestore-chat",
    "platforms": [
      "ios",
      "android",
      "web"
    ],
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true
    },
    "description": "POC for Firestore and React Native",
    "githubUrl": "https://github.com/apappas1129/firestore-chat",
    "android": {
      "package": "com.deleonandroid3.firestorechat"
    }
  }
}
