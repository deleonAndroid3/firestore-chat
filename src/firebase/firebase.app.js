/* eslint-disable no-undef */
import firebase from 'firebase'
import '@firebase/auth'
import '@firebase/firestore'

const firebaseConfig = {
  apiKey: 'AIzaSyCI8XZTr0uYVfsOekGf9HyibI8oTAD_HTk',
  authDomain: 'reactnativefirebase-78631.firebaseapp.com',
  databaseURL: 'https://reactnativefirebase-78631.firebaseio.com',
  projectId: 'reactnativefirebase-78631',
  storageBucket: 'reactnativefirebase-78631.appspot.com',
  messagingSenderId: '243790233194',
  appId: '1:243790233194:web:44e3e983f7bdb587c39413',
}

if (!firebase.apps.length) {
  firebase.initializeApp(firebaseConfig)
}

const firestore = firebase.firestore()

export { firebase, firestore }
