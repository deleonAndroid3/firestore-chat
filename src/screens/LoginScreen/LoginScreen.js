import React, { useState } from 'react'
import { CheckBox, Image, Text, TextInput, TouchableOpacity, View } from 'react-native'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import styles from './styles'
import { firebase } from '../../firebase/firebase.app'
import * as Facebook from 'expo-facebook'
import env from '../../../environment'

/**FOR Unit Testing */
export function handleSignIn(email, password) {
  return firebase
    .auth()
    .signInWithEmailAndPassword(email, password)
}

export function handleSignOut() {
  if (firebase.auth().currentUser) {
    return firebase.auth().signOut()
  }
}

export async function handleSignInWithCred(_type, _credential){
  
  try {
    /**
     * If the user or Facebook cancelled the login, returns { type: 'cancel' }.
     * Otherwise, returns { type: 'success' } & [FacebookAuthenticationCredential]
     * 
     * let us assume what the user does using the _type parameter  
     */
    // await Facebook.initializeAsync({
    //   appId: '821584115048039',
    // })
    // const { type, token } = await Facebook.logInWithReadPermissionsAsync({
    //   permissions: ['public_profile', 'email'],
    // })
    if (_type === 'success') {
      // const credential = firebase.auth.FacebookAuthProvider.credential(token)
      return firebase
        .auth()
        .signInWithCredential(_credential)
    }else{
      return false
    }

  }catch (error){
    return error
  }
}
/**End Of Unit Testing */


export default function LoginScreen({ navigation }) {
  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')

  const onFooterLinkPress = () => {
    navigation.navigate('Registration')
  }
  
  const onLoginPress = () => {
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then((response) => {
        const uid = response.user.uid
        const usersRef = firebase.firestore().collection('users')
        usersRef
          .doc(uid)
          .get()
          .then((firestoreDocument) => {
            if (!firestoreDocument.exists) {
              alert('User does not exist anymore.')
              return
            }
            // const user = firestoreDocument.data()
            // navigation.navigate('Home', { user: user })
          })
          .catch((error) => {
            alert(error)
          })
      })
      .catch((error) => {
        alert(error)
      })
  }

  async function loginWithFacebook() {
    try {
      await Facebook.initializeAsync({
        appId: env.facebookAppID,
      })
      const { type, token } = await Facebook.logInWithReadPermissionsAsync({
        permissions: ['public_profile', 'email'],
      })
      if (type === 'success') {
        const credential = firebase.auth.FacebookAuthProvider.credential(token)

        firebase
          .auth()
          .signInWithCredential(credential)
          .then(({ additionalUserInfo, user }) => {
            const { profile, providerId } = additionalUserInfo
            const data = {
              _id: user.uid,
              email: profile.email || user.email || user.providerData[0].email,
              name: profile.name || user.displayName,
              avatar:
                profile.picture?.data?.url ||
                user.photoURL ||
                user.providerData[0].photoURL,
              providerId,
              photo: `https://graph.facebook.com/${profile.id}/picture?height=500`,
            }
            const usersRef = firebase.firestore().collection('users')
            usersRef
              .doc(user.uid)
              .set(data)
              .then(() => {
                // navigation.navigate('Home', { user: data })
              })
              .catch((error) => {
                alert(error)
              })
          })
          .catch((error) => {
            alert(error)
          })
      }
    } catch (error) {
      console.warn('FB_LOGIN_FAILED', JSON.stringify(error))
      alert(`Facebook Login Error: ${error.message}`)
    }
  }

  return (
    <View style={styles.container}>
      <KeyboardAwareScrollView
        style={{ flex: 1, width: '100%' }}
        keyboardShouldPersistTaps='always'
      >
        <Image
          style={styles.logo}
          source={require('../../../assets/icon.png')}
        />
        <TextInput
          style={styles.input}
          placeholder='E-mail'
          placeholderTextColor='#aaaaaa'
          onChangeText={(text) => setEmail(text)}
          value={email}
          underlineColorAndroid='transparent'
          autoCapitalize='none'
          testID='Email'
          accessibilityLabel='Email'
        />
        <TextInput
          style={styles.input}
          placeholderTextColor='#aaaaaa'
          secureTextEntry
          placeholder='Password'
          onChangeText={(text) => setPassword(text)}
          value={password}
          underlineColorAndroid='transparent'
          autoCapitalize='none'
          testID='Password'
          accessibilityLabel='Password'
        />
        <TouchableOpacity style={styles.button} onPress={() => onLoginPress()} testID='Login' accessibilityLabel='Login'>
          <Text style={styles.buttonTitle}>Log in</Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.button} onPress={() => loginWithFacebook()}testID='LoginFB' accessibilityLabel='LoginFB'>
          <Text style={styles.buttonTitle}>Log in With Facebook</Text>
        </TouchableOpacity>

        <View style={styles.footerView}>
          <Text style={styles.footerText}>
            Don&apos;t have an account?{' '}
            <Text onPress={onFooterLinkPress} style={styles.footerLink}>
              Sign up
            </Text>
          </Text>
        </View>
      </KeyboardAwareScrollView>
    </View>
  )
}
