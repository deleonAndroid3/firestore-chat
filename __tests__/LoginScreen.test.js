/* eslint-disable no-undef */
import { shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-test-renderer'
import LoginScreen, { signin } from '../src/screens/LoginScreen/LoginScreen'
import { firebase } from '../src/firebase/firebase.app'


describe('< LoginScreen />', () => {

  const user = {
    _id: '93nJbIsNNRMezRIXoIgKUg9PKh42',
    email: 'apappas1129@gmail.com',
    name: 'Ts Xs',
  }

  firebase.auth = jest.fn().mockReturnValue({
    currentUser: {
      displayName: 'testDisplayName',
      email: 'test@test.com',
      emailVerified: true
    },
    signOut: function() { return true },
    signInWithEmailAndPassword: function(email, password) { 
      if (email === 'apappas1129@gmail.com' && password === 'password') {
        return user 
      }else{
        return 'user not found'
      }
    },
  })

  let login =  shallow(<LoginScreen />)
 
  it('App renders without crashing', async () => {
    await act(async () => {
      expect(login).toBeTruthy()
    })
  })

  it('App test against snapshot', async () => {
    await act(async () => {
      expect(login).toMatchSnapshot()
    })
  })

  it('App should render components', () => {
    expect(login.find('Component[placeholder="E-mail"]').exists()).toBe(true)
    expect(login.find('Component[placeholder="Password"]').exists()).toBe(true)
    expect(login.find('ForwardRef').exists()).toBe(true)
  })

  it('should return user', () => {
    let result = signin('apappas1129@gmail.com', 'password')
    expect(result).toEqual(user)
  })

  it('should return error', () => {
    let result = signin('apappas1129@gmail.com', 'notapassword')
    expect(result).toEqual('user not found')
  })
})
