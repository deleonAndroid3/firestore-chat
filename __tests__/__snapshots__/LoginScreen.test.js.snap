// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< LoginScreen /> App test against snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <KeyboardAwareScrollView
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="always"
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
    viewIsInsideTabBar={false}
  >
    <Image
      source={1}
      style={
        Object {
          "alignSelf": "center",
          "flex": 1,
          "height": 120,
          "margin": 30,
          "width": 90,
        }
      }
    />
    <Component
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="E-mail"
      placeholderTextColor="#aaaaaa"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "height": 48,
          "marginBottom": 10,
          "marginLeft": 30,
          "marginRight": 30,
          "marginTop": 10,
          "overflow": "hidden",
          "paddingLeft": 16,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <Component
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="#aaaaaa"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "height": 48,
          "marginBottom": 10,
          "marginLeft": 30,
          "marginRight": 30,
          "marginTop": 10,
          "overflow": "hidden",
          "paddingLeft": 16,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <ForwardRef
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#788eec",
          "borderRadius": 5,
          "height": 48,
          "justifyContent": "center",
          "marginLeft": 30,
          "marginRight": 30,
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Log in
      </Text>
    </ForwardRef>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#2e2e2d",
            "fontSize": 16,
          }
        }
      >
        Don't have an account?
         
        <Text
          onPress={[Function]}
          style={
            Object {
              "color": "#788eec",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Sign up
        </Text>
      </Text>
    </View>
  </KeyboardAwareScrollView>
</View>
`;
